File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.ArrayList;
import java.util.HashMap;
class Solution {
    public static String[] solution(String[] record) {
        ArrayList<String> ment = new ArrayList<>();
        HashMap<String, String> idMap = new HashMap<>();
        String[] answer;
        String[] keyward;
         for (String cmd : record) {            keyward = cmd.split(" ");
            if (keyward[0].equals("Enter") || keyward[0].equals("Change"))
                idMap.put(keyward[1], keyward[2]);
        }
        for (String cmd : record) {
            keyward = cmd.split(" ");
            if (keyward[0].equals("Enter"))
                ment.add(idMap.get(keyward[1]) + "님이 들어왔습니다.");
            else if (keyward[0].equals("Leave"))
                ment.add(idMap.get(keyward[1]) + "님이 나갔습니다.");
        }
        answer = ment.toArray(new String[0]);
        return answer;
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <sstream>
#include <unordered_map>

using namespace std;
//개선

vector<string> solution(vector<string> record) {
    vector<string> answer;
    vector<pair<string, string>> logs;
    unordered_map<string, string> nameMap;

    for (const string& log : record)
    {
        stringstream ss(log);
        string cmd, ID, nickName;
        ss >> cmd >> ID >> nickName;

        //닉네임이 변경되는 경우
        if ("Enter" == cmd || "Change" == cmd) { nameMap[ID] = nickName; }

        //메시지에 기록해야하는 경우
        if ("Enter" == cmd || "Leave" == cmd) { logs.push_back({ cmd, ID }); ; }
    }

    for (const auto& log2 : logs)
    {
        string msg = nameMap[log2.second] + "님이 ";
        msg += (log2.first == "Enter") ? "들어왔습니다." : "나갔습니다.";
        answer.push_back(msg);
    }

    return answer;
}